File: Python3/1457_pseudo_palindromic_paths_in_a_binary_tree.py
# https://leetcode.cn/problems/pseudo-palindromic-paths-in-a-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        cnts = [0 for _ in range(10)]

        def dfs(node):
            if node is None:
                return 0
            cnts[node.val] = 1 - cnts[node.val]
            if node.left is None and node.right is None:
                if sum(cnts) <= 1:
                    res = 1
                else:
                    res = 0
            else:
                res = dfs(node.left) + dfs(node.right)
            cnts[node.val] = 1 - cnts[node.val]
            return res
            
        return dfs(root)

File: Cpp/1457_pseudo_palindromic_paths_in_a_binary_tree.cpp
// https://leetcode.cn/problems/pseudo-palindromic-paths-in-a-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pseudoPalindromicPaths (TreeNode* root) {
        vector<int> cnts(10, 0);
        return dfs(root, cnts);
    }

    int dfs(TreeNode *node, vector<int> &cnts) {
        if (node == nullptr) {
            return 0;
        }
        cnts[node->val] = 1 - cnts[node->val];
        int res = 0;
        if (node->left == nullptr && node->right == nullptr) {
            int s = 0;
            for (auto cnt: cnts) {
                s += cnt;
            }
            if (s <= 1) {
                res = 1;
            }
        }
        else {
            res = dfs(node->left, cnts) + dfs(node->right, cnts);
        }
        cnts[node->val] = 1 - cnts[node->val];

        return res;
    }
};